# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: gfx/wr/webrender/src/prim_store/mod.rs
# Commit: ea22ccce0c4d
# Full Hash: ea22ccce0c4dbe8a1c7ace807924275497f6c899
# Author: Glenn Watson <git@intuitionlibrary.com>
# Date: 2022-05-25 09:36:24
# Regressor Bug: 1769963
# File Overlap Count: 1
# Description:
#   Bug 1769963 - Fix backdrop-filter inside complex transform with CSS filter chain r=gfx-reviewers,lsalzman
#   
#   Differential Revision: https://phabricator.services.mozilla.com/D147124
# ==============================================================================

diff -r 60c64f026b6b -r ea22ccce0c4d gfx/wr/webrender/src/prim_store/mod.rs
--- a/gfx/wr/webrender/src/prim_store/mod.rs	Wed May 25 05:48:08 2022 +0300
+++ b/gfx/wr/webrender/src/prim_store/mod.rs	Wed May 25 02:59:32 2022 +0000
@@ -28,7 +28,7 @@
 #[cfg(debug_assertions)]
 use std::sync::atomic::{AtomicUsize, Ordering};
 use crate::util::Recycler;
-use crate::internal_types::LayoutPrimitiveInfo;
+use crate::internal_types::{FastHashSet, LayoutPrimitiveInfo};
 #[cfg(debug_assertions)]
 use crate::internal_types::FrameId;
 use crate::visibility::PrimitiveVisibility;
@@ -44,7 +44,7 @@
 
 mod storage;
 
-use backdrop::BackdropDataHandle;
+use backdrop::{BackdropCaptureDataHandle, BackdropRenderDataHandle};
 use borders::{ImageBorderDataHandle, NormalBorderDataHandle};
 use gradient::{LinearGradientPrimitive, LinearGradientDataHandle, RadialGradientDataHandle, ConicGradientDataHandle};
 use image::{ImageDataHandle, ImageInstance, YuvImageDataHandle};
@@ -124,6 +124,10 @@
 #[cfg_attr(feature = "replay", derive(Deserialize))]
 pub struct PictureIndex(pub usize);
 
+impl PictureIndex {
+    pub const INVALID: PictureIndex = PictureIndex(!0);
+}
+
 #[cfg_attr(feature = "capture", derive(Serialize))]
 #[cfg_attr(feature = "replay", derive(Deserialize))]
 #[derive(Copy, Debug, Clone, MallocSizeOf, PartialEq)]
@@ -1060,8 +1064,12 @@
         data_handle: PrimitiveDataHandle,
     },
     /// Render a portion of a specified backdrop.
-    Backdrop {
-        data_handle: BackdropDataHandle,
+    BackdropCapture {
+        data_handle: BackdropCaptureDataHandle,
+    },
+    BackdropRender {
+        data_handle: BackdropRenderDataHandle,
+        pic_index: PictureIndex,
     },
 }
 
@@ -1192,7 +1200,10 @@
             PrimitiveInstanceKind::YuvImage { data_handle, .. } => {
                 data_handle.uid()
             }
-            PrimitiveInstanceKind::Backdrop { data_handle, .. } => {
+            PrimitiveInstanceKind::BackdropCapture { data_handle, .. } => {
+                data_handle.uid()
+            }
+            PrimitiveInstanceKind::BackdropRender { data_handle, .. } => {
                 data_handle.uid()
             }
         }
@@ -1257,6 +1268,9 @@
 
     /// List of current debug messages to log on screen
     messages: Vec<DebugMessage>,
+
+    /// Set of sub-graphs that are required, determined during visibility pass
+    pub required_sub_graphs: FastHashSet<PictureIndex>,
 }
 
 impl Default for PrimitiveScratchBuffer {
@@ -1270,6 +1284,7 @@
             gradient_tiles: GradientTileStorage::new(0),
             debug_items: Vec::new(),
             messages: Vec::new(),
+            required_sub_graphs: FastHashSet::default(),
         }
     }
 }
@@ -1300,6 +1315,8 @@
         //           should fix this in the future to retain handles.
         self.gradient_tiles.clear();
 
+        self.required_sub_graphs.clear();
+
         self.debug_items.clear();
     }
 